python/cuml/ts/arima.pyx

```diff
@@ -0,0 +1,814 @@
+        d_vs, d_vs_ptr, _, _, _ = input_to_dev_array(vs, check_dtype=np.float64)
+
+        cdef uintptr_t d_y_ptr    
+        if self.d_y is None:
+            d_y, d_y_ptr, _, _, _ = input_to_dev_array(self.y, check_dtype=np.float64)
+            self.d_y = (d_y, d_y_ptr)
+        
+        (_, d_y_ptr) = self.d_y
+
+        predict_in_sample(handle_[0], <double*>d_y_ptr,
+                          self.num_batches, self.num_samples, p, d, q,
+                          <double*>d_params_ptr, <double*>d_vs_ptr, <double*>d_y_p_ptr)
+
+        
+        
+        updateHost(&vs[0,0], <double*>d_vs_ptr, (self.num_samples-1) * self.num_batches, 0)
+        updateHost(&y_p[0,0], <double*>d_y_p_ptr, (self.num_samples) * self.num_batches, 0)
+
+        self.yp = y_p
+        return y_p
+
+    def forecast(self, nsteps: int) -> np.ndarray:
+        """Forecast the given model `nsteps` into the future.
+
+        Parameters:
+        ----------
+        nsteps : int
+                 The number of steps to forecast beyond end of `fit()` signal
+
+        Returns:
+        --------
+        y_fc : array-like, shape = (nsteps, n_series)
+               Forecast, one for each series.
+
+        Example:
+        --------
+        .. code-block:: python
+            from cuml.ts.arima import fit
+            ...
+            model = fit(ys, (1,1,1), mu0, ar0, ma0)
+            y_fc = model.forecast(10)
+            dx = xs[1] - xs[0]
+            xfc = np.linspace(1, 1+50*dx, 50)
+            plt.plot(xs, ys, xfc, yfc)
+
+        """
+
+        p, d, q = self.order[0]
+        
+        handle = cuml.common.handle.Handle()
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>handle.getHandle()
+
+        x = pack(p, d, q, self.num_batches, self.mu, self.ar_params, self.ma_params)
+        cdef uintptr_t d_params_ptr
+        d_params, d_params_ptr, _, _, _ = input_to_dev_array(x, check_dtype=np.float64)
+
+        cdef np.ndarray[double, ndim=2, mode="fortran"] y_p = np.zeros(((self.num_samples),
+                                                                        self.num_batches), order="F")
+        cdef uintptr_t d_y_p_ptr
+        d_y_p, d_y_p_ptr, _, _, _ = input_to_dev_array(y_p, check_dtype=np.float64)
+
+        cdef np.ndarray[double, ndim=2, mode="fortran"] vs = np.zeros(((self.num_samples - d),
+                                                                       self.num_batches), order="F")
+        cdef uintptr_t d_vs_ptr
+        d_vs, d_vs_ptr, _, _, _ = input_to_dev_array(vs, check_dtype=np.float64)
+
+        cdef uintptr_t d_y_ptr    
+        if self.d_y is None:
+            d_y, d_y_ptr, _, _, _ = input_to_dev_array(self.y, check_dtype=np.float64)
+            self.d_y = (d_y, d_y_ptr)
+            
+        (_, d_y_ptr) = self.d_y
+
+        residual(handle_[0], <double*>d_y_ptr, self.num_batches, self.num_samples, p, d, q,
+                 <double*>d_params_ptr, <double*>d_vs_ptr,
+                 False)
```

Comment:
Same comments as `predict_in_sample` apply here, by storing device arrays in init this can be simplified and we can save time transferring to device at object creation instead of method calling